### PR TITLE
Fix issue 20059 - mismatched function return type inference should gi…

### PR DESCRIPTION
**Pull request by thewilsonator**
…ve location of inferred type

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: src/dmd/statementsem.d [hunk 1]
```diff
@@ -3225,7 +3225,12 @@ else
                     }
                     else if (rs.exp.op != TOK.error)
                     {
-                        rs.error("mismatched function return type inference of `%s` and `%s`", rs.exp.type.toChars(), tret.toChars());
+                        Loc prevretloc = (!fd.returns) ? fd.loc : (*fd.returns)[0].loc;
```
Comment (JinShil) [inline src/dmd/statementsem.d]: Variable names should be camelCase according to [the D Style](https://dlang.org/dstyle.html#naming_general)

Comment (JinShil) [reply src/dmd/statementsem.d]: Also I think it's more idiomatic to use `auto` for the storage class in this case.

Comment (thewilsonator) [reply src/dmd/statementsem.d]: Fixed.

File: test/fail_compilation/diag20059.d [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/diag20059.d(14): Error: mismatched function return type inference of `string[]` and `string` (inferred at fail_compilation/diag20059.d(12))
```
Comment (ZombineDev) [inline test/fail_compilation/diag20059.d]: Use `errorSupplemental`, to make the error message a bit easier to digest:

```suggestion
fail_compilation/diag20059.d(14): Error: Expected return type of `string`, but got `string[]`.
fail_compilation/diag20059.d(12):    Return type of `string` inferred here.
```

Comment (thewilsonator) [reply test/fail_compilation/diag20059.d]: Done.

**Review by PetarKirov — APPROVED** [review]
Thanks!

**Pull request merged**